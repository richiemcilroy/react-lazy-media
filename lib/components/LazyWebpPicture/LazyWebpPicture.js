'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
exports.LazyWebpPicture = void 0;

var _react = _interopRequireWildcard(require('react'));

var _propTypes = _interopRequireDefault(require('prop-types'));

var _ReactLazyload = _interopRequireDefault(require('../../ReactLazyload'));

var _LazyImage = require('../LazyImage/LazyImage');

function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : {default: obj};
}

function _interopRequireWildcard(obj) {
  if (obj && obj.__esModule) {
    return obj;
  } else {
    var newObj = {};
    if (obj != null) {
      for (var key in obj) {
        if (Object.prototype.hasOwnProperty.call(obj, key)) {
          var desc =
            Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {};
          if (desc.get || desc.set) {
            Object.defineProperty(newObj, key, desc);
          } else {
            newObj[key] = obj[key];
          }
        }
      }
    }
    newObj['default'] = obj;
    return newObj;
  }
}

function _typeof(obj) {
  if (typeof Symbol === 'function' && typeof Symbol.iterator === 'symbol') {
    _typeof = function _typeof(obj) {
      return typeof obj;
    };
  } else {
    _typeof = function _typeof(obj) {
      return obj && typeof Symbol === 'function' && obj.constructor === Symbol && obj !== Symbol.prototype
        ? 'symbol'
        : typeof obj;
    };
  }
  return _typeof(obj);
}

function _classCallCheck(instance, Constructor) {
  if (!(instance instanceof Constructor)) {
    throw new TypeError('Cannot call a class as a function');
  }
}

function _defineProperties(target, props) {
  for (var i = 0; i < props.length; i++) {
    var descriptor = props[i];
    descriptor.enumerable = descriptor.enumerable || false;
    descriptor.configurable = true;
    if ('value' in descriptor) descriptor.writable = true;
    Object.defineProperty(target, descriptor.key, descriptor);
  }
}

function _createClass(Constructor, protoProps, staticProps) {
  if (protoProps) _defineProperties(Constructor.prototype, protoProps);
  if (staticProps) _defineProperties(Constructor, staticProps);
  return Constructor;
}

function _possibleConstructorReturn(self, call) {
  if (call && (_typeof(call) === 'object' || typeof call === 'function')) {
    return call;
  }
  return _assertThisInitialized(self);
}

function _assertThisInitialized(self) {
  if (self === void 0) {
    throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  }
  return self;
}

function _getPrototypeOf(o) {
  _getPrototypeOf = Object.setPrototypeOf
    ? Object.getPrototypeOf
    : function _getPrototypeOf(o) {
        return o.__proto__ || Object.getPrototypeOf(o);
      };
  return _getPrototypeOf(o);
}

function _inherits(subClass, superClass) {
  if (typeof superClass !== 'function' && superClass !== null) {
    throw new TypeError('Super expression must either be null or a function');
  }
  subClass.prototype = Object.create(superClass && superClass.prototype, {
    constructor: {value: subClass, writable: true, configurable: true}
  });
  if (superClass) _setPrototypeOf(subClass, superClass);
}

function _setPrototypeOf(o, p) {
  _setPrototypeOf =
    Object.setPrototypeOf ||
    function _setPrototypeOf(o, p) {
      o.__proto__ = p;
      return o;
    };
  return _setPrototypeOf(o, p);
}

function _defineProperty(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {value: value, enumerable: true, configurable: true, writable: true});
  } else {
    obj[key] = value;
  }
  return obj;
}

var LazyWebpPicture =
  /*#__PURE__*/
  (function(_Component) {
    _inherits(LazyWebpPicture, _Component);

    function LazyWebpPicture() {
      _classCallCheck(this, LazyWebpPicture);

      return _possibleConstructorReturn(this, _getPrototypeOf(LazyWebpPicture).apply(this, arguments));
    }

    _createClass(LazyWebpPicture, [
      {
        key: 'componentDidMount',
        // Update lazyLoad after first rendering of every image
        value: function componentDidMount() {
          _ReactLazyload['default'].getInstance().update();
        } // Update lazyLoad after re-rendering of every image
      },
      {
        key: 'componentDidUpdate',
        value: function componentDidUpdate() {
          _ReactLazyload['default'].getInstance().update();
        } // TODO - Don't try to load the webp image if the file does not exist.
      },
      {
        key: 'render',
        value: function render() {
          var webp = this.props.webp || ''.concat(this.props.src, '.webp');
          return _react['default'].createElement(
            'picture',
            null,
            _react['default'].createElement('source', {
              type: 'image/webp',
              'data-srcset': webp
            }),
            _react['default'].createElement(_LazyImage.LazyImage, this.props)
          );
        }
      }
    ]);

    return LazyWebpPicture;
  })(_react.Component);

exports.LazyWebpPicture = LazyWebpPicture;

_defineProperty(LazyWebpPicture, 'propTypes', {
  alt: _propTypes['default'].string,
  classes: _propTypes['default'].arrayOf(_propTypes['default'].string),
  id: _propTypes['default'].string,
  src: _propTypes['default'].string.isRequired,
  width: _propTypes['default'].number,
  height: _propTypes['default'].number,
  webp: _propTypes['default'].string,
  lqip: _propTypes['default'].string // Low-quality image placeholder
});

_defineProperty(LazyWebpPicture, 'defaultProps', {
  classes: []
});
