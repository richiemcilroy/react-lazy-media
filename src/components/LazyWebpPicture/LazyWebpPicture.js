import React, {Component} from 'react';
import PropTypes from 'prop-types';
import ReactLazyload from '../../ReactLazyload';
import {LazyImage} from '../LazyImage/LazyImage';

export class LazyWebpPicture extends Component {
  static propTypes = {
    alt: PropTypes.string,
    classes: PropTypes.arrayOf(PropTypes.string),
    id: PropTypes.string,
    src: PropTypes.string.isRequired,
    width: PropTypes.number,
    height: PropTypes.number,
    webp: PropTypes.string,
    lqip: PropTypes.string // Low-quality image placeholder
  };

  static defaultProps = {
    classes: []
  };

  // Update lazyLoad after first rendering of every image
  componentDidMount() {
    ReactLazyload.getInstance().update();
  }

  // Update lazyLoad after re-rendering of every image
  componentDidUpdate() {
    ReactLazyload.getInstance().update();
  }

  // TODO - Don't try to load the webp image if the file does not exist.
  render() {
    const webp = this.props.webp || `${this.props.src}.webp`;

    return (
      <picture>
        <source type="image/webp" data-srcset={webp} />
        <LazyImage {...this.props} />
      </picture>
    );
  }
}
