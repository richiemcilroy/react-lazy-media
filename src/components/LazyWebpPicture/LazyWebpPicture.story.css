.lazyWebpPicture {
  width: 530px;
  height: 600px;
  margin-top: 150vh;
}
