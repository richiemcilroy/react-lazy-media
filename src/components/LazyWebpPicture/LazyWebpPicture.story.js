import React from 'react';
import {storiesOf} from '@storybook/react';
import {withKnobs, text} from '@storybook/addon-knobs';
import {LazyWebpPicture} from './LazyWebpPicture';
import './LazyWebpPicture.story.css';
import png from './Hero.png';
import webp from './Hero.png.webp';
import lqip from './LQIP.jpeg';

storiesOf('Lazy', module)
  .addDecorator(withKnobs)
  .add('LazyWebpPicture', () => {
    return <LazyWebpPicture classes={['lazyWebpPicture']} src={png} webp={webp} lqip={lqip} />;
  });
