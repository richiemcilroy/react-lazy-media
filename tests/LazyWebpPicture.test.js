import React from 'react';
import renderer from 'react-test-renderer';
import {LazyWebpPicture} from '../src/components/LazyWebpPicture/LazyWebpPicture';

test('LazyWebpPicture should require a src prop', () => {
  expect(() => {
    renderer.create(<LazyWebpPicture />);
  }).toThrowErrorMatchingSnapshot();
});

test('LazyWebpPicture should have default props and children', () => {
  const component = renderer.create(<LazyWebpPicture src={'/image.jpeg'} />);

  const tree = component.toJSON();

  const srcChild = tree.children.shift();

  expect(srcChild.type).toBe('source');

  expect(srcChild.props).toEqual({
    'data-srcset': '/image.jpeg.webp',
    type: 'image/webp'
  });

  const lazyImageChild = tree.children.shift();

  expect(lazyImageChild.type).toBe('img');

  expect(lazyImageChild.props).toMatchObject({
    'data-src': '/image.jpeg',
    className: 'lazy'
  });
});

test('LazyWebpPicture should render an picture with a src', () => {
  const component = renderer.create(<LazyWebpPicture src={'/image.jpeg'} />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('LazyWebpPicture should render a picture with a src and specified webp src', () => {
  const component = renderer.create(<LazyWebpPicture src={'/image.jpeg'} webp={'/image.webp'} />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
