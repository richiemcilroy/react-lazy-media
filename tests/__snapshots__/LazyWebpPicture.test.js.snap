// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LazyWebpPicture should render an picture with a src 1`] = `
<picture>
  <source
    data-srcset="/image.jpeg.webp"
    type="image/webp"
  />
  <img
    className="lazy"
    data-src="/image.jpeg"
  />
</picture>
`;

exports[`LazyWebpPicture should render an picture with a webp src 1`] = `
<picture>
  <source
    data-srcset="/image.webp"
    type="image/webp"
  />
  <img
    className="lazy"
    data-src="/image.jpeg"
  />
</picture>
`;

exports[`LazyWebpPicture should require a src prop 1`] = `
"Warning: Failed prop type: The prop \`src\` is marked as required in \`LazyWebpPicture\`, but its value is \`undefined\`.
    in LazyWebpPicture"
`;
